### PR TITLE
feat: build concurrency, unsafe-perm, scriptsPrependNodePath (#397 items 12, 14, 15)

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Bundles the three remaining moderate/minor parity items from #397 into one PR. Each adds a `pnpm-workspace.yaml` setting, threads it through `Config`, and applies it to `BuildModules` so a pnpm yaml that exercises these knobs produces the same install behavior in pacquet.

- **Item 12 — `childConcurrency`** ([upstream `getWorkspaceConcurrency`](https://github.com/pnpm/pnpm/blob/b4f8f47ac2/config/reader/src/concurrency.ts#L25-L34); [`runGroups(...)`](https://github.com/pnpm/pnpm/blob/b4f8f47ac2/building/during-install/src/index.ts#L124)). Adds `Config.child_concurrency: u32` with the upstream-matching default `min(4, availableParallelism())` and the negative-offset semantics (`n < 0` → `max(1, parallelism - |n|)`). `BuildModules::run` now dispatches each chunk's members across a bounded rayon thread pool via `par_iter().try_for_each` — chunks themselves remain sequential to preserve topological order. `ignored_builds` and `deps_state_cache` are wrapped in `Mutex` so the recursive memo and the dedup set survive concurrent chunk-member access.
- **Item 14 — `unsafePerm`**. Adds `Config.unsafe_perm: bool` (default `true`) and threads it through to `RunPostinstallHooks`. When `false`, the executor sets `TMPDIR=node_modules/.tmp` for the spawn. The uid/gid drop side is a no-op upstream too because pnpm's [`@pnpm/npm-lifecycle`](https://github.com/pnpm/npm-lifecycle/blob/d2d8e790/index.js#L204-L220) never populates `opts.user` / `opts.group`. The POSIX auto-detect (`getuid() === 0 && setgid → unsafePerm = false`) needs `libc`, which isn't in `[workspace.dependencies]` yet — for now, root-run CI must set `unsafePerm: false` in yaml explicitly. Windows is force-overridden to `true` in `WorkspaceSettings::apply_to`, matching upstream's `process.platform === 'win32'` gate.
- **Item 15 — `scriptsPrependNodePath`** ([`Config.scriptsPrependNodePath`](https://github.com/pnpm/pnpm/blob/b4f8f47ac2/config/reader/src/Config.ts#L108)). Adds a tri-state `ScriptsPrependNodePath` enum in `pacquet-config` (custom serde `Deserialize` for the `boolean | "warn-only"` yaml shape) and converts to `pacquet_executor::ScriptsPrependNodePath` at the `BuildModules` call site so the executor crate stays free of serde wiring. Default is `Never` to match upstream's [`StrictBuildOptions.scriptsPrependNodePath: false`](https://github.com/pnpm/pnpm/blob/b4f8f47ac2/building/after-install/src/extendBuildOptions.ts#L78).
- **Item 16 — `getSubgraphToBuild` filter trimming**. Already implemented in pacquet's `get_subgraph_to_build` via the `child_should_be_built || needs_build || has_patch` gate. Confirmed during this slice; no code change needed.

Why one PR: items 12 / 14 / 15 are three independent yaml-driven config knobs, all small. Bundling keeps the cross-crate signature churn from happening three times (`BuildModules`'s field list, `install_frozen_lockfile`'s call site, and the test fixtures all touch each one), and the upstream sources for the three settings live next to each other so the porting context is shared.

## Test plan

- [x] `cargo nextest run` — all 605 workspace tests pass.
- [x] `just ready` — typos, fmt, check, lint clean.
- [x] `cargo doc -D warnings --document-private-items` — clean.
- [x] `taplo format --check` — clean.
- [x] `parses_scripts_prepend_node_path_{true,false,warn_only}_from_yaml`, `rejects_invalid_scripts_prepend_node_path`.
- [x] `parses_unsafe_perm_from_yaml_and_applies`, `unsafe_perm_force_true_on_windows` (Windows-only).
- [x] `parses_{positive,negative}_child_concurrency_from_yaml_and_*`.
- [x] Existing `BuildModules` tests cover the concurrency-1 path (so the rayon refactor doesn't regress the single-threaded case).

---
Written by an agent (Claude Code, claude-opus-4-7).

## Summary by CodeRabbit

* **New Features**
  * Workspace options: scriptsPrependNodePath (tri‑state), unsafePerm, and childConcurrency; childConcurrency now defaults sensibly based on CPU.
  * Lifecycle scripts run with configurable parallelism; non‑positive childConcurrency supports “use all cores minus N”.

* **Behavior**
  * Windows safely defaults unsafePerm when unspecified.
  * Deterministic build ordering and improved side‑effects cache/skip reporting; thread‑pool failures now surfaced.

* **Tests**
  * Added YAML parsing, concurrency edge cases, platform behavior, and build-sequence tests.

* **Documentation**
  * Clarified scripts-prepend behavior in executor docs.

[Review Change Stack](https://app.coderabbit.ai/change-stack/pnpm/pacquet/pull/429)

### CODE REVIEW

File: crates/config/src/defaults.rs [hunk 1]
```diff
@@ -105,5 +105,42 @@ pub fn default_fetch_retry_maxtimeout() -> u64 {
...
+/// - Zero or negative `n` → `max(1, parallelism - |n|)`.
+///
+/// The negative-offset semantics let users say "use all cores minus
+/// N" without hardcoding the core count.
+pub fn resolve_child_concurrency(option: Option<i32>) -> u32 {
+    match option {
+        None => default_child_concurrency(),
+        Some(n) if n > 0 => n as u32,
+        Some(n) => {
+            let offset = (-n) as u32;
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/config/src/defaults.rs]: Fixed in 026ba284 — replaced `(-n) as u32` with `n.unsigned_abs()` in `resolve_child_concurrency`. Added `resolve_child_concurrency_handles_i32_min` to lock in the non-panicking floor.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/config/src/defaults.rs [hunk 2]
```diff
@@ -105,5 +105,42 @@ pub fn default_fetch_retry_maxtimeout() -> u64 {
...
+            available_parallelism().saturating_sub(offset).max(1)
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/config/src/defaults.rs]: Fixed in 026ba284. Same underlying issue as the Copilot comment on the next line — replaced `(-n) as u32` with `n.unsigned_abs()` and added `resolve_child_concurrency_handles_i32_min` to lock it in.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/config/src/lib.rs [hunk 3]
```diff
@@ -39,6 +40,57 @@ pub enum NodeLinker {
     Pnp,
 }
 
+/// Tri-state mirror of `pacquet_executor::ScriptsPrependNodePath`
+/// with serde wiring. The executor crate keeps its own enum free of
+/// serde so config concerns don't leak into the spawn-path. Converted
+/// at the BuildModules call site via [`From`].
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/config/src/lib.rs]: Fixed in 026ba284 — corrected the doc to reflect what the code actually does. Adding a real `From` impl would require either `pacquet-config` to depend on `pacquet-executor` or the reverse, which would invert the layering for one three-arm conversion that already lives at the `BuildModules` call site. Doc now describes the explicit `match` and explains why no `From` exists.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/package-manager/src/build_modules.rs [hunk 4]
```diff
@@ -301,279 +326,345 @@ impl<'a> BuildModules<'a> {
...
+        // process's rayon-global threads (which other crates may
+        // depend on). One pool reused across all chunks; chunks
+        // themselves run sequentially.
+        //
+        // Mirrors upstream's
+        // [`runGroups(getWorkspaceConcurrency(opts.childConcurrency), groups)`](https://github.com/pnpm/pnpm/blob/b4f8f47ac2/building/during-install/src/index.ts#L124).
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(child_concurrency.max(1) as usize)
+            .build()
+            .expect("build the rayon pool for the chunk member dispatch");
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/build_modules.rs]: Fixed in 026ba284 — `ThreadPoolBuilder::build()` now propagates via a new `BuildModulesError::ThreadPoolBuild` variant (`ERR_PACQUET_BUILD_THREAD_POOL`) carrying the underlying `rayon::ThreadPoolBuildError` and a remediation hint about lowering `childConcurrency` / raising `RLIMIT_NPROC`. Callers see a structured error instead of a panic.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/package-manager/src/build_modules.rs [hunk 5]
```diff
@@ -301,279 +326,345 @@ impl<'a> BuildModules<'a> {
...
+            }
+            None => {
+                // "Not in allowBuilds" — surfaced as
+                // `pnpm:ignored-scripts`. Explicit `false` is
+                // silently denied (above), matching upstream's
+                // switch.
+                ignored_builds
+                    .lock()
+                    .expect("ignored_builds mutex")
+                    .insert(metadata_key.to_string());
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/build_modules.rs:505]: Fixed in 026ba284 — every `Mutex::lock().expect(...)` and `into_inner().expect(...)` around `ignored_builds` and `deps_state_cache` is now `unwrap_or_else(|e| e.into_inner())`. Recovery is safe here: the protected types are `BTreeSet<String>` and `DepsStateCache` (HashMap-backed memo), and each worker performs one atomic insertion per call — a panic mid-call cannot leave the structure in a torn state. Each call site carries a comment with the safety argument.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/config/src/defaults/tests.rs [hunk 6]
```diff
@@ -46,6 +46,18 @@ fn test_default_store_dir_with_xdg_env() {
...
+/// A naive `(-n) as u32` overflows in debug builds when
+/// `n == i32::MIN` because the negation itself overflows;
+/// `unsigned_abs` is the safe path. Bounds the result against
+/// the same `[1, parallelism]` envelope as the negative-offset
+/// path normally produces.
+#[test]
+fn resolve_child_concurrency_handles_i32_min() {
+    let result = resolve_child_concurrency(Some(i32::MIN));
+    assert!(result >= 1, "must floor at 1 even for i32::MIN");
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/config/src/defaults/tests.rs:140]: Fixed in 1af1a813 — tightened to `assert_eq!(result, 1)`. The exact result is provably `1` (`i32::MIN.unsigned_abs()` is `2_147_483_648`, larger than any plausible host parallelism, so `saturating_sub` produces 0 and `.max(1)` lifts to 1).

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/package-manager/src/build_modules.rs [hunk 7]
```diff
@@ -40,6 +42,19 @@ pub enum BuildModulesError {
...
+    /// (`EAGAIN` / RLIMIT_NPROC). Surfaced as a structured error
+    /// rather than a panic so the install path can return cleanly.
+    #[display("Failed to build the per-install rayon thread pool: {source}")]
+    #[diagnostic(
+        code(ERR_PACQUET_BUILD_THREAD_POOL),
+        help(
+            "Lower childConcurrency in pnpm-workspace.yaml, or raise the process's RLIMIT_NPROC."
+        )
+    )]
+    ThreadPoolBuild { source: rayon::ThreadPoolBuildError },
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/build_modules.rs]: Fixed in ce8bce03 — added `#[error(source)]` on the `source` field. Confirmed the convention is consistent across the workspace (e.g. `StoreIndexError::*`).

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/package-manager/src/build_sequence.rs [hunk 8]
```diff
@@ -141,6 +141,18 @@ fn collect_root_dep_paths(
...
+    // a second root B sharing that child gets trimmed). Upstream's
+    // input arrives in JS-object insertion order; pacquet sources
+    // these from `HashMap<_, ProjectSnapshot>` and
+    // `HashMap<PkgName, ResolvedDependencySpec>`, so iteration order
+    // is non-deterministic. Sort by `PackageKey` string repr so the
+    // build sequence (and the trim behavior) is reproducible run to
+    // run. Long-term fix is to preserve lockfile declaration order
+    // via `IndexMap`; until then, an alphabetical sort is enough to
+    // make the build path deterministic.
+    roots.sort_by_key(|k| k.to_string());
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/build_sequence.rs:162]: Fixed in 90dbdbf1 — applied the suggested `child_keys.sort_by_key(|k| k.to_string())` in `build_children_map` so transitive shared descendants are also walked in a deterministic order. Comment links the safety argument back to the matching sort in `collect_root_dep_paths`.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/package-manager/src/build_modules.rs [hunk 9]
```diff
@@ -301,279 +342,367 @@ impl<'a> BuildModules<'a> {
...
-                }
+            // The closure runs once per chunk; `try_for_each`
+            // short-circuits on the first error. The only mutable
+            // state shared across tasks is the two `Mutex`-wrapped
+            // collections above and `deps_state_cache`.
+            pool.install(|| -> Result<(), BuildModulesError> {
+                chunk.par_iter().try_for_each(|snapshot_key| {
+                    build_one_snapshot::<R>(
+                        snapshot_key,
+                        snapshots,
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/build_modules.rs:391]: Fixed in e5dcb8a6 — added `build_modules_collects_ignored_builds_under_concurrency` which mirrors the sequential `build_modules_collects_ignored_builds` fixture but runs with `child_concurrency: 2`. Two policy-denied build candidates land in chunk 0 (no dep edges between them) so the rayon `pool.install` / `par_iter().try_for_each` dispatch is actually exercised. The `BTreeSet`-backed `ignored_builds` normalizes the concurrent insertion order, so the expected output is the same sorted set as the sequential variant — a regression that dropped the `Mutex` would surface as a rustc/clippy error rather than a runtime failure, but the test at least documents the parallel codepath is reachable.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/package-manager/src/build_modules/tests.rs [hunk 10]
```diff
@@ -311,6 +316,75 @@ fn build_modules_collects_ignored_builds() {
...
+        patches: None,
+
+        scripts_prepend_node_path: ScriptsPrependNodePath::Never,
+        unsafe_perm: true,
+        child_concurrency: 2,
+    }
+    .run::<SilentReporter>()
+    .expect("run BuildModules under concurrency");
+    dbg!(&ignored);
+
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/build_modules/tests.rs:377]: Keeping the `dbg!` per pacquet's documented test-logging convention in [`CODE_STYLE_GUIDE.md`](https://github.com/pnpm/pacquet/blob/main/CODE_STYLE_GUIDE.md):

> If the values being compared with `assert_eq!` have complex structures (such as a struct or an array), they should be logged with `dbg!`.

`ignored` is `Vec<String>`, which qualifies as an array under that rule, and the sibling tests (`build_modules_collects_ignored_builds`, `build_modules_excludes_explicit_deny_from_ignored`, `do_not_fail_on_optional_dep_with_failing_postinstall`) all use the same `dbg!(&ignored);` pattern. The harness captures debug output on successful runs, so it won't show in CI logs unless the test fails. Removing it here would diverge from the established convention.

---
Written by an agent (Claude Code, claude-opus-4-7).
